src/filebrowser/transfer-mgr.cpp

```diff
@@ -123,19 +379,831 @@ void TransferManager::cancelAllDownloadTasks()
+    sqlite_foreach_selected_row(db_, zql, getFileTasksByIdCB, &file_tasks_);
+    sqlite3_free(zql);
+
+    zql = sqlite3_mprintf("SELECT * FROM FolderTasks WHERE status = 'pending' ");
+    sqlite_foreach_selected_row(db_, zql, getFolderTasksByIdCB, &folder_tasks_);
+    sqlite3_free(zql);
+
+    // update total_bytes of folder tasks
+    QMap<int, FolderTaskRecord>::const_iterator ite = folder_tasks_.constBegin();
+    while (ite != folder_tasks_.constEnd()) {
+        const int folder_task_id = ite.key();
+        QMap<int, FileTaskRecord>::const_iterator file_ite = file_tasks_.constBegin();
+        while (file_ite != file_tasks_.constEnd()) {
+            if (file_ite.value().parent_task == folder_task_id) {
+                folder_tasks_[folder_task_id].total_bytes +=
+                    QFileInfo(file_ite.value().local_path).size();
+            }
+            ++file_ite;
+        }
+        ++ite;
+    }
+}
+
+const FileTaskRecord *TransferManager::getPendingFileTask()
+{
+    QMap <int, FileTaskRecord>::const_iterator ite =
+        file_tasks_.constBegin();
+    while (ite != file_tasks_.constEnd()) {
+        if (ite.value().status == kPendingStatus) {
+            return &ite.value();
+        }
+        ++ite;
+    }
+    return NULL;
+}
+
+void TransferManager::addUploadTask(const QString& repo_id,
+                                    const QString& repo_parent_path,
+                                    const QString& local_path,
+                                    const QString& name,
+                                    int folder_task_id)
+{
+    QString repo_full_path = ::pathJoin(repo_parent_path, name);
+
+    if (isCurrentUploadTask(repo_id, repo_full_path, local_path) ||
+        getUploadTaskId(repo_id, repo_full_path, local_path)) {
+        return;
+    }
+
+    QString task_type;
+    if (QFileInfo(local_path).isDir()) {
+        task_type = "dir";
+    } else {
+        task_type = "file";
+    }
+
+    if (db_) {
+        char *zql = NULL;
+        if (!folder_task_id) {
+            zql = sqlite3_mprintf(
+                "REPLACE INTO FileTasks (account_signature, "
+                "type, repo_id, path, local_path, status ) "
+                "VALUES (%Q, %Q, %Q, %Q, %Q, 'pending') ",
+                toCStr(account_signature_),
+                toCStr(task_type),
+                toCStr(repo_id),
+                toCStr(repo_full_path),
+                toCStr(local_path));
+        } else {
+             zql = sqlite3_mprintf(
+                "REPLACE INTO FileTasks (account_signature, "
+                "type, repo_id, path, local_path, status, parent_task ) "
+                "VALUES (%Q, %Q, %Q, %Q, %Q, 'pending', %Q) ",
+                toCStr(account_signature_),
+                toCStr(task_type),
+                toCStr(repo_id),
+                toCStr(repo_full_path),
+                toCStr(local_path),
+                toCStr(QString::number(folder_task_id)));
```

Comment:
Change `%Q` to `%d`, then you don't need `toCStr(QString::number(folder_task_id)))`